### PR TITLE
[PyTorch] Make RecordFunction store inputs as ArrayRef

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76266
* #72551
* __->__ #72484

Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

### CODE REVIEW

File: torch/csrc/autograd/function.h [hunk 1]
```diff
@@ -160,9 +160,10 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
         // the forward pass function
         guard.setForwardThreadId(thread_id_);
         if (guard.needsInputs()) {
+          std::vector<c10::IValue> inputs_vec(inputs.begin(), inputs.end());
           guard.before(
             name(),
-            std::vector<c10::IValue>(inputs.begin(), inputs.end()),
+            c10::ArrayRef<const c10::IValue>(inputs_vec.data(), inputs_vec.size()),
```
Comment (albanD) [inline torch/csrc/autograd/function.h:166]: Is that ok that the inputs_vec is invalidated just after this call and before anything actually runs?

Comment (swolchok) [reply torch/csrc/autograd/function.h]: yes -- we debug-check that. before() calls start callbacks, which are the only things that should be touching inputs.

**Review by robieta — APPROVED** [review]

**Comment by swolchok** [comment]
failures look unrelated, landing

**Pull request closed without merging**